Moment/ImageMoments.tcc

```diff
@@ -0,0 +1,695 @@
+
+        // Need to initialize this way rather than doing it in the constructor, because using a single Array in the constructor means that
+        // all Arrays in the vector reference the same Array.
+        for (casacore::uInt k = 0; k < n_out; k++) {
+            result_arrays[k] = Array<T>(result_array_shape);
+            result_array_masks[k] = Array<Bool>(result_array_shape);
+        }
+
+        // Iterate through a chunk, slice by slice on the output image display axes
+        casacore::Bool done = casacore::False;
+        while (!done) {
+            if (_stop) {
+                break;
+            }
+
+            casacore::Vector<T> data(chunk(chunk_slice_start, chunk_slice_end));
+            casacore::Vector<Bool> mask =
+                use_mask ? casacore::Vector<casacore::Bool>(mask_chunk(chunk_slice_start, chunk_slice_end)) : no_mask;
+            cur_pos = iter_pos + chunk_slice_start;
+
+            // Do calculations
+            collapser.multiProcess(result, result_mask, data, mask, cur_pos);
+
+            // Fill partial results in a chunk
+            for (uInt k = 0; k < n_out; ++k) {
+                result_arrays[k](chunk_slice_start) = result[k];
+                result_array_masks[k](chunk_slice_start) = result_mask[k];
+            }
+
+            done = True; // The scan of this chunk is complete
+
+            // Proceed to the next slice on the display axes
+            for (casacore::uInt k = 0; k < n_display_axes; ++k) {
+                casacore::uInt dax = display_axes[k];
+                if (chunk_slice_start[dax] < chunk_shape[dax] - 1) {
+                    ++chunk_slice_start[dax];
+                    ++chunk_slice_end[dax];
+                    done = False;
+                    break;
+                } else {
+                    chunk_slice_start[dax] = 0;
+                    chunk_slice_end[dax] = 0;
+                }
+            }
+        }
+
+        // Put partial results in the output lattices (as a chunk size)
+        for (casacore::uInt k = 0; k < n_out; ++k) {
+            casacore::IPosition result_pos = in_ndim == out_dim ? iter_pos : iter_pos.removeAxes(casacore::IPosition(1, collapse_axis));
+            casacore::Bool keep_axis = result_arrays[k].ndim() == lattice_out[k]->ndim();
+            if (!keep_axis) {
+                result_arrays[k].removeDegenerate(display_axes);
+            }
+            lattice_out[k]->putSlice(result_arrays[k], result_pos);
+
+            if (lattice_out[k]->hasPixelMask()) {
+                casacore::Lattice<casacore::Bool>& mask_out = lattice_out[k]->pixelMask();
+                if (mask_out.isWritable()) {
+                    if (!keep_axis) {
+                        result_array_masks[k].removeDegenerate(display_axes);
+                    }
+                    mask_out.putSlice(result_array_masks[k], result_pos);
+                }
+            }
+        }
+
+        if (tell_progress != 0) {
```

Comment:
why don't we move the `if(_stop)` check to here? So that it's consistent with the `tell_progress` callback